[WIP] Two nodes openshift (#1675)
openshift/enhancements: enhancements/two-nodes-openshift/2no_template.md

Given these factors, a two-node deployment of OpenShift offers a consistent, reliable solution that meets the needs of customers across all their sites, from central management hubs to the most remote edge locations.

### User Stories

* As a large enterprise with multiple remote sites, I want a cost-effective OpenShift cluster solution so that I can manage containers without the overhead of a third node.
* As a support engineer, I want an automated method for handling the failure of a single node so that I can quickly restore service and maintain system integrity.
* As an infrastructure administrator, I want to ensure seamless failover for virtual machines (VMs) so that in the event of a node failure, the VMs are automatically migrated to a healthy node with minimal downtime and no data loss.
* As a network operator, I want my Cloud-Native Network Functions (CNFs) to be orchestrated consistently using OpenShift, regardless of whether they are in datacenters, or at the far edge where physical space is limited.

### Goals

> * Implement a highly available two-node OpenShift cluster.

Reviewer: We might want to specifically call out here (or somewhere else in the enhancement) the number and types of failures we can maintain availability for.

Author: We do have a section about [failure handling](https://github.com/openshift/enhancements/pull/1675/commits/b0d63a6e3a76bb30d34fe98f8dfb030b9b05bf29#diff-0339498404cfe988e6d38c75d19b6e12a7b502fdecdc66c0469fc6408843aea8R120), my initial thought was if it's not in that section we don't support it.
Maybe it's enough to just be more clear about it and clearly express that notion in the enhancement or did you have something else in mind ?

Reviewer: I'd like us be more precise with the wording.  "highly available" means lots of things to different people.

Reviewer: Resolved

   3. Stop failure is optionally escalated to a node failure (fencing)
   4. Start failure defaults to leaving the service offline
6. Etcd Failure
   1. Pacemaker’s monitoring detects the failure
   2. Pacemaker either demotes etcd so it can resync, or restarts and promotes etcd
   3. Stop failure is optionally escalated to a node failure (fencing)
   4. Start failure defaults to leaving the service offline

### API Extensions

API Extensions are CRDs, admission and conversion webhooks, aggregated API servers,
and finalizers, i.e. those mechanisms that change the OCP API surface and behaviour.

- Name the API extensions this enhancement adds or modifies.
- Does this enhancement modify the behaviour of existing resources, especially those owned
  by other parties than the authoring team (including upstream resources), and, if yes, how?
  Please add those other parties as reviewers to the enhancement.

  Examples:
  - Adds a finalizer to namespaces. Namespace cannot be deleted without our controller running.
  - Restricts the label format for objects to X.
  - Defaults field Y on object kind Z.

Fill in the operational impact of these API Extensions in the "Operational Aspects

> of API Extensions" section.

Author: >something that can signal CEO 

I think that we prefer something "general" which can be used by other stakeholder process (instead of creating multiple localized flags).

> of API Extensions" section.

Reviewer 2: What about introducing a feature flag for this? Can be read by all components which need to deal with 2NO, and can be set in install config AFAIK.

> of API Extensions" section.

Reviewer 3: we need to signal _after_ the installation is completed

> of API Extensions" section.

Author: > we need to signal after the installation is completed

Hmm I'm not sure that's an issue, can't we turn on the flag as part of the installation process ? (i.e after the originally bootstrapped node has joined the cluster as a regular node)

> of API Extensions" section.

Reviewer 3: You guys need to signal after all your stuff is ready to run etcd, is there a guarantee that this is true after the bootstrap node has joined the cluster?

I was under the impression there is a fair bit of installation involved.

> of API Extensions" section.

Author: >is there a guarantee that this is true after the bootstrap node has joined the cluster as a real control plane node?

The way I understand it: past that point we can set up pacemaker/corosync (so it would be after the node joining the cluster).
IIUC in this particular context it's not very important where (timing wise) is the point in which we can signal CEO to relinquish control of etcd, as to the fact that there is such a point in time (in which we'll turn the flag).

> of API Extensions" section.

Reviewer 3: That's something you should write down in this enhancement, in which case you're accepting some amount of downtime of the API during the installation. Not sure if there's also race conditions involved where podman is coming up faster than the static etcd pod and we're in an eternal static pod revision rollout crash loop.

Note that a bunch of CI jobs will not going to like this kind of disruption.

> of API Extensions" section.

Author: Doing a short recap following our last sync:
- Since we want CEO to relinquish etcd control to pacemaker after bootstrapping is complete we can't relay on the installer to do that.
- One potential option is using LCO (Local Cluster Operator) to somehow flag CEO to relinquish control of etcd.
- What kind of flag we will use to trigger MCO is still TBD

Let me know if that makes sense.

> of API Extensions" section.

Reviewer: Resolved

#### Hypershift / Hosted Control Planes

Are there any unique considerations for making this change work with
Hypershift?

See https://github.com/openshift/enhancements/blob/e044f84e9b2bafa600e6c24e35d226463c2308a5/enhancements/multi-arch/heterogeneous-architecture-clusters.md?plain=1#L282

How does it affect any of the components running in the
management cluster? How does it affect any components running split
between the management cluster and guest cluster?

#### Standalone Clusters

Is the change relevant for standalone clusters?

#### Single-node Deployments or MicroShift

How does this proposal affect the resource consumption of a
single-node OpenShift deployment (SNO), CPU and memory?

How does this proposal affect MicroShift? For example, if the proposal
adds configuration options through API resources, should any of those
behaviors also be exposed to MicroShift admins through the
configuration file for MicroShift?

### Implementation Details/Notes/Constraints

#### Installation flow
1. We’ll set up Pacemaker and Corosync on RHCOS using MCO layering.
   * [TBD extend more]
2. Install an “SNO like” first node using a second bootstrapped node.

>    * This is somewhat similar what is done in SNO CI in AWS (up until the part the bootstrapped node is removed) and is possible because CEO can distinguish the bootstrapped node as a special use case, thus enabling its removal without breaking the etcd quorum for the remaining node.

Reviewer: IMHO it would be more beneficial to be specific about what we're doing rather than explaining by analogy.
eg.

1. Deploy bootstrap node to machine A
2. Do thing to machine B
3. ...

Author: Thanks, makes sense.

@rwsu if you can help with phrasing  a line or two to better phrase it from the installer perspective, that would be great

Reviewer: Resolved

   We should be safe after [MGMT-13586](https://issues.redhat.com/browse/MGMT-13586) which makes the installer wait for the bootstrap etcd member to be removed first before shutting it down.
3. After the bootstrapped node is removed add it to the cluster as a “regular” node.
4. Switch CEO to “2NO” mode (where it does not manage etcd) and remove the etcd static pods
   * [TBD localized/global switch]
   * This is done because we want to allow simpler maintenance and keeping  some of CEO functionality (defragmentation, cert rotation ect…)
5. Configure Pacemaker/Corosync to manage etcd/kubelet/cri-o
6. [TBD storage]

#### Fencing / quorum management

> Fencing quorum is managed by corosync and etcd  will be managed by Pacemaker which will force the etcd quorum.

Reviewer: ```suggestion
Fencing and quorum is managed by corosync and etcd  will be managed by Pacemaker which will force the etcd quorum.
```

> Fencing quorum is managed by corosync and etcd  will be managed by Pacemaker which will force the etcd quorum.

Reviewer: I'm confused by this sentence. Do you mean "fencing quorum" or "fencing and quorum".
The former would be a new term that's not been defined, and the second would be incorrect (corosync does not manage fencing).

Suggest breaking this into two independent sentences

Author: ```suggestion
Corosync will manage the quorum.
Pacemaker will apply node fencing and etcd management (promote/demote) according to Corosyncs quorum
```
I've tried to rephrase, let me know if that works

Reviewer: There's two types of quorum.

Maybe: Corosync will provide an independent definition of quorum

Reviewer: Resolved

> Fencing quorum is managed by corosync and etcd  will be managed by Pacemaker which will force the etcd quorum.

Author: ```suggestion
Corosync will manage the quorum.
Pacemaker will apply node fencing and etcd management (promote/demote) according to Corosyncs quorum.

Here is a node failure example demonstrating that:
1. Corosync on the survivor (Node1)
2. Etcd loses internal quorum (E-quorum) and goes read-only
3. Node1 retains “corosync quorum” (C-quorum) and initiates fencing of Node2
4. Once fencing is successful Pacemaker will use a fence/resource agent (TBD) which will reschedule the workload from the fenced node
5. Pacemaker on Node1 forces E-quorum (etcd promotion event)
6. Cluster continues with no redundancy
7. … time passes …
8. Node2 has a persistent failure that prevents communication with Node1
   * Node2 does not have C-quorum (requires forming a membership with it’s peer)
   * Node2 does not start etcd or kubelet, remains inert waiting for Node1
9. Persistent failure on Node2 is repaired
10. Corosync membership containing both nodes forms
11. Pacemaker “starts” etcd on Node2 as a follower of Node1
12. Pacemaker “promotes” etcd on Node2 as full replica of Node1
13. Pacemaker starts kubelet
14. Cluster continues with 1+1 redundancy

[Here](#failure-handling) is a  more extensive list of failure scenarios.

#### CEO Enhancement

> 1. Requires a new infrastructure type in OpenShift APIs

Reviewer: We need to articulate why.

Author: ```suggestion
1. Requires a new infrastructure type in OpenShift APIs mainly for two purposes:
   1. Signaling/Indicating CEO runs in 2NO mode
   2. Signal CEO when to relinquish control over etcd management 
```
@tjungblu please keep me honest here and correct me in case I got it wrong.

Reviewer 3: the infrastructure CRD is more general than CEO and is used by any consumer of the platform to determine what and where it is running.
I don't think we can solve (2) through the infrastructure CRD alone

Author: >I don't think we can solve (2) through the infrastructure CRD alone

I think that's fine, the intention is just that it's part of the solution.

Reviewer: Resolved

> 2. Replicate the functionality of CEO that is tied to static pods? e.g. Certificate rotation, backup/restore, apiserver etcd endpoint controller

Reviewer 4: This say's you're planning to replicate this functionality, but https://github.com/openshift/enhancements/pull/1675/files#diff-0339498404cfe988e6d38c75d19b6e12a7b502fdecdc66c0469fc6408843aea8R248-R250 say's you're building a special CEO mode to retain the ability for it to run this functionality, which is it?

Author: ```suggestion
2. Make sure that even though CEO will not manage etcd, it will still retain other relevant capabilities (defragmentation, certificate rotation, backup/restore etc...).

Author: >which is it?

Good comment, I've tried to better rephrase it.

> 3. Some functionality to “know” when to switch to “2NO mode”

Reviewer: This section doesn't feel complete

Author: ATM we're still discussing it [here](https://github.com/openshift/enhancements/pull/1675/files#r1761047077) , I'll make sure to update this section once we reach a conclusion.

Reviewer: Resolved

### Risks and Mitigations

#### Risks:

> 1. On a failed node scenario ,The node with elected Pacemaker DC (designated coordinator) will fence the other node. That brings up the question of how to schedule the workload of the fenced nodes ? [mitigation](#scheduling-workload-on-fenced-nodes)

Reviewer: ```suggestion
1. In the event of a node failure, Pacemaker on the survivor will fence the other node and cause etcd to recover quorum. However this will not automatically recover affected workloads. [mitigation](#scheduling-workload-on-fenced-nodes)
```

However I think this is part of the design, not a risk to be mitigated.  Risks might happen, this will happen on every node failure.

Author: Yep. good comment.
However resolving this is crucial for a working 2NO cluster, so maybe an open question with some [alternative solutions ](https://github.com/openshift/enhancements/pull/1675/files#diff-0339498404cfe988e6d38c75d19b6e12a7b502fdecdc66c0469fc6408843aea8R301-R308)is the way to go here ?

Reviewer: Resolved

2. We plan to configure Pacemaker to manage etcd and give it quorum (for example after the remaining node fence its peer in a failed node use case)  [mitigation](#pacemaker-controlling-key-elements)
   1. How do we plan Pacemaker to give the etcd quorum and which consideration should be taken ?

>    2. How does Pacemaker giving etcd quorum affect other etcd stakeholders (etcd pod, etcd operator, etc…)  ?

Reviewer: What's the risk here?  An open question is different from a risk.

Instead of "How do we do X?" type questions, try to think in terms of "If we do (or can't do) X then {risk}".

Reviewer: And the open questions, should be listed in the section at line ~388

Reviewer: Resolved

* Defragmentation
* Certificate creation and rotation
* Active etcd endpoint export for apiserver (etcd-endpoints configmap in openshift-config namespace)
* Installation of the Backup/Restore scripts

Source as of 4.15: [CEO <> CEE](https://docs.google.com/presentation/d/1U_IyNGHCAZFAZXyzAs5XybR8qT91QaQ2wr3W9w9pSaw/edit#slide=id.g184d8fd7fc3_1_99)

### Drawbacks

The idea is to find the best form of an argument why this enhancement should
_not_ be implemented.

What trade-offs (technical/efficiency cost, user experience, flexibility,
supportability, etc) must be made in order to implement this? What are the reasons
we might not want to undertake this proposal, and how do we overcome them?

Does this proposal implement a behavior that's new/unique/novel? Is it poorly
aligned with existing user expectations?  Will it be a significant maintenance
burden?  Is it likely to be superceded by something else in the near future?

## Open Questions [optional]

This is where to call out areas of the design that require closure before deciding
to implement the design.  For instance,
 > 1. This requires exposing previously private resources which contain sensitive

>   information.  Can we do this?

Author: ```suggestion
1. If we are giving Pacemaker control over etcd then other etcd stakeholders (etcd pod, etcd operator, etc…) might be affected.

Reviewer: Resolved
